### PR TITLE
Feature:support configuring PodManagementPolicy in TidbCluster CR

### PR DESCRIPTION
**Pull request by mianhk**
### What problem does this PR solve?

Close: #3228 
### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Feature: support configuring PodManagementPolicy in TidbCluster CR
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/tidbcluster_component.go [hunk 1]
```diff
@@ -101,6 +103,24 @@ func (a *componentAccessorImpl) StatefulSetUpdateStrategy() apps.StatefulSetUpda
...
+		policy = a.ComponentSpec.PodManagementPolicy
+	}
+
+	// unified podManagementPolicy check to avoid check everywhere
+	if policy == apps.OrderedReadyPodManagement {
+		return apps.OrderedReadyPodManagement
+	}
+	return apps.ParallelPodManagement
+}
+
```
Comment (KanShiori) [inline pkg/apis/pingcap/v1alpha1/tidbcluster_component.go:123]: How about initializing `policy` as ParallelPodManagement?
```
func (a *componentAccessorImpl) PodManagementPolicy() apps.PodManagementPolicyType {
	policy := apps.ParallelPodManagement
        // modify policy
        
        // don't check 
        return policy
```

File: pkg/apis/pingcap/v1alpha1/tidbcluster_component.go [hunk 2]
```diff
@@ -101,6 +103,24 @@ func (a *componentAccessorImpl) StatefulSetUpdateStrategy() apps.StatefulSetUpda
 	return a.ComponentSpec.StatefulSetUpdateStrategy
 }
 
+func (a *componentAccessorImpl) PodManagementPolicy() apps.PodManagementPolicyType {
+	var policy apps.PodManagementPolicyType
+	if a.ComponentSpec == nil || len(a.ComponentSpec.PodManagementPolicy) == 0 {
```
Comment (july2993) [inline pkg/apis/pingcap/v1alpha1/tidbcluster_component.go:108]: suggest change: 
```go
if a.ComponentSpec != nil && len(a.ComponentSpec.PodManagementPolicy) != 0 { // handle case by case with highest priority first
  policy = a.ComponentSpec.PodManagementPolicy
else if len(a.podManagementPolicy) != 0{
  policy = a.podManagementPolicy
}

// handle the final value we get `policy`
....
```

File: examples/advanced/tidb-cluster.yaml [hunk 3]
```diff
@@ -134,6 +134,10 @@ spec:
   ## If specified, the pods will be scheduled by the specified scheduler.
   ## Can be overwritten by component settings.
   schedulerName: tidb-scheduler
+  
+  ## PodManagementPolicy default Parallel.
+  ## Ref: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#pod-management-policies
+  podManagementPolicy: Parallel
```
Comment (july2993) [inline examples/advanced/tidb-cluster.yaml:140]: Do we need to change the default value as OrderedReady?
@DanielZhangQD

**Review by july2993 — COMMENTED** [review]
rest LGTM

File: pkg/apis/pingcap/v1alpha1/tidbcluster_component.go [hunk 4]
```diff
@@ -101,6 +103,24 @@ func (a *componentAccessorImpl) StatefulSetUpdateStrategy() apps.StatefulSetUpda
...
+		}
+		policy = a.podManagementPolicy
+	} else {
+		policy = a.ComponentSpec.PodManagementPolicy
+	}
+
+	// unified podManagementPolicy check to avoid check everywhere
+	if policy == apps.OrderedReadyPodManagement {
+		return apps.OrderedReadyPodManagement
+	}
```
[the remark this answers is not in the document]
Comment (mianhk) [reply pkg/apis/pingcap/v1alpha1/tidbcluster_component.go:120]: Yes, Refer to StatefulSetUpdateStrategy： 
https://github.com/pingcap/tidb-operator/blob/48ae8298f491c0f7c57ef53041912cffcb5ce6a6/pkg/manager/member/pd_member_manager.go#L759-L767

File: pkg/apis/pingcap/v1alpha1/tidbcluster_component.go (hunk 2 above)
Comment (mianhk) [reply pkg/apis/pingcap/v1alpha1/tidbcluster_component.go]: Optimized in the following commit, THX.

File: examples/advanced/tidb-cluster.yaml (hunk 3 above)
Comment (DanielZhangQD) [reply examples/advanced/tidb-cluster.yaml:140]: We need to keep the current behavior for the default configuration.

Comment (july2993) [reply examples/advanced/tidb-cluster.yaml:140]: ok, for pump it is not `keep the current behavior for the default configuration`, it does not set the value before

File: pkg/manager/member/pump_member_manager.go [hunk 5]
```diff
@@ -470,6 +470,7 @@ func getNewPumpStatefulSet(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap) (*app
 
 			Template:             podTemplate,
 			VolumeClaimTemplates: volumeClaims,
+			PodManagementPolicy:  spec.PodManagementPolicy(),
```
Comment (DanielZhangQD) [inline pkg/manager/member/pump_member_manager.go:473]: The `podManagementPolicy` is not set before as @july2993 has said in another comment, so the default value is `OrderedReady` for pump, and this field is not allowed to update after the sts is created, so we have to set to `OrderedReady` for pump if it's not set explicitly.

File: pkg/manager/member/pump_member_manager.go [hunk 6]
```diff
@@ -461,6 +461,18 @@ func getNewPumpStatefulSet(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap) (*app
...
+	if len(tc.Spec.Pump.PodManagementPolicy) != 0 {
+		podManagementPolicy = tc.Spec.Pump.PodManagementPolicy
+	} else if len(tc.Spec.PodManagementPolicy) != 0 {
+		podManagementPolicy = tc.Spec.PodManagementPolicy
+	}
+
+	if podManagementPolicy != apps.OrderedReadyPodManagement && podManagementPolicy != apps.ParallelPodManagement {
+		podManagementPolicy = apps.OrderedReadyPodManagement
+	}
+
```
Comment (DanielZhangQD) [inline pkg/manager/member/pump_member_manager.go:475]:
```suggestion
	if len(tc.Spec.Pump.PodManagementPolicy) != 0 || len(tc.Spec.PodManagementPolicy) != 0 {
		podManagementPolicy = BasePumpSpec.PodManagementPolicy()
	}

```

File: pkg/manager/member/pump_member_manager.go [hunk 7]
```diff
@@ -461,6 +461,12 @@ func getNewPumpStatefulSet(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap) (*app
 		Spec: podSpec,
 	}
 
+	// To compatible with default podManagementPolicy of pump is "OrderedReady"
+	podManagementPolicy := apps.OrderedReadyPodManagement
+	if len(tc.Spec.Pump.PodManagementPolicy) != 0 || len(tc.Spec.PodManagementPolicy) != 0 {
+		podManagementPolicy = BasePumpSpec.PodManagementPolicy()
```
Comment (DanielZhangQD) [inline pkg/manager/member/pump_member_manager.go:467]:
```suggestion
		podManagementPolicy = spec.PodManagementPolicy()
```

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**